packages/core/content-manager/server/src/history/services/history.ts

```diff
@@ -349,24 +330,115 @@ const createHistoryService = ({ strapi }: { strapi: Core.Strapi }) => {
+                  currentRelationData.results.push(relatedEntry);
+                } else {
+                  // The related content has been deleted
+                  currentRelationData.meta.missingCount += 1;
+                }
+
+                return currentRelationData;
+              },
+              Promise.resolve({
+                results: [] as any[],
+                meta: { missingCount: 0 },
+              })
+            )
+        );
+      };
+
       const populateEntryRelations = async (
         entry: HistoryVersionQueryResult
       ): Promise<CreateHistoryVersion['data']> => {
         const entryWithRelations = await Object.entries(entry.schema).reduce(
           async (currentDataWithRelations, [attributeKey, attributeSchema]) => {
+            const attributeValue = entry.data[attributeKey];
+            const attributeValues = Array.isArray(attributeValue)
+              ? attributeValue
+              : [attributeValue];
+
+            if (attributeSchema.type === 'media') {
+              return {
+                ...(await currentDataWithRelations),
+                [attributeKey]: await buildMediaResponse(attributeValues),
+              };
+            }
+
             // TODO: handle relations that are inside components
-            if (['relation', 'media'].includes(attributeSchema.type)) {
-              const attributeValue = entry.data[attributeKey];
-              const relationResponse = await buildRelationReponse(
-                (Array.isArray(attributeValue)
-                  ? attributeValue
-                  : [attributeValue]) as EntryToPopulate[],
-                attributeSchema
-              );
+            if (
+              attributeSchema.type === 'relation' &&
+              attributeSchema.relation !== 'morphToOne' &&
+              attributeSchema.relation !== 'morphToMany'
+            ) {
+              /**
+               * Don't build the relations response object for relations to admin users,
+               * because pickAllowedAdminUserFields sanitizati
+               */
+              if (attributeSchema.target === 'admin::user') {
+                const sanitizedAdminUser = await Promise.all(
+                  attributeValues.map(async (userToPopulate) => {
+                    if (userToPopulate == null) {
+                      return null;
+                    }
+
+                    const user = await strapi
+                      .query('admin::user')
+                      .findOne({ where: { id: userToPopulate.id } });
+
+                    const permissionChecker = getContentManagerService('permission-checker').create(
+                      {
+                        userAbility: params.state.userAbility,
+                        model: 'admin::user',
+                      }
+                    );
+                    const sanitizedUser = await permissionChecker.sanitizeOutput(user);
+
+                    return sanitizedUser;
```

Comment:
can't you sanitize at the end? I mean, sanitizing the entire entry at once

Reply:
just saw remis comment , he was asking the same

Reply:
the one about moving the user sanitization logic outside the loop

Reply:
ahh I had understood that he wanted me to move the creation of the permission checker outside of the loop since it only needs to be created once. Is that what you meant here as well?